### DISCUSSION TITLE
Shader language needs to provide support for line width

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#6455] [3D]
I'm trying to draw some guide lines in a program.  These are being generated with an ImmediateMesh object are a set of lines created with Mesh.PRIMITIVE_LINE_STRIP.  At the moment, they are being drawn at 1 pixel thickness and there seems to be no way to change this.  The StandardMaterial3D node provides no support for this, and even the shading language seems to have no line size support:

https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/spatial_shader.html

I tried using POINT_SIZE, but that does not affect my lines.

Setting line size is a standard operation in OpenGL, and it would be useful to see it supported by Godot's shaders as well.

**Comment by AThousandShips** [comment]
Is it universally supported by Vulkan though?

For OpenGL even the line width isn't, or at least wasn't, universally supported

Also the ranges of widths that are supported vary massively, some drivers only support 1-64, 0-8, 0.5-10, 0-8000, and some only 1 (some AMD Radeon)

**Comment by blackears** [comment]
For my purposes, I just need to make the lines slightly thicker to create a visual distinction.  I don't know how Vulkan would handle this, but even if it was just 2 or 3 pixels and wasn't universally supported, it would still be helpful.

**Reply by AThousandShips** [reply]
So what happens with your game if it's run on hardware that doesn't support 3 px thick lines?

**Reply by AThousandShips** [reply]
If we integrate this we're gonna get a lot of issues being opened by people who say their game doesn't run right on some hardware or that the feature doesn't work right, because of how unevenly it is implemented, this is why the editor uses other methods to draw lines instead of relying on this feature (which has had bugs in the past)

**Comment by blackears** [comment]
It's not for the game, it's for an editor plugin.  And as I would expect the number of platforms that do not support it to be fairly small, I'd just accept the degraded performance on those platforms.  Or you could just add a some sort of call to OS that tells you the capabilities of the graphics card.

How does the editor implement this feature?

**Reply by AThousandShips** [reply]
But it's not just a matter of having the feature or not, as I said

What should the editor do if the hardware only supports lines of width 1? Or only up to 7.99219? Fall back to a hard coded one?

The editor uses custom line rendering the same as the game does, but adding it *only* for the editor makes no sense, if it's for a plugin you can just write your own wide line rendering, or just use the wide lines that are built in

**Reply by AThousandShips** [reply]
Also the platforms that don't *really* support it isn't fairly small, as I said, and at least given gpuinfo it's essentially unsupported on MacOS, though that might be limited information

According to gpuinfo:
0% support on MacOS
17% of the hardware that supports it only supports 1 px lines, though a large portion of that is Android

See also #2490

**Reply by blackears** [reply]
How does the editor handle it?  Draw camera facing quads?  Extrude a cylinder along the line path?

**Reply by AThousandShips** [reply]
I'm not sure for 3D sorry, but in 2D at least it uses triangle strips, wasn't clear that you desired 3D sorry

**Comment by Calinou** [comment]
> Setting line size is a standard operation in OpenGL, and it would be useful to see it supported by Godot's shaders as well.

Only NVIDIA supports line widths greater than 1. AMD and Intel are limited to 1 (at least on Windows).

This poor support is why Vulkan doesn't feature a line width setting and expects you to draw thick lines using polygons instead.
